js/render/draw_extrusion.js

```diff
@@ -0,0 +1,292 @@
+'use strict';
+
+var mat3 = require('gl-matrix').mat3;
+var mat4 = require('gl-matrix').mat4;
+var vec3 = require('gl-matrix').vec3;
+var pixelsToTileUnits = require('../source/pixels_to_tile_units');
+var Buffer = require('../data/buffer');
+var VertexArrayObject = require('./vertex_array_object');
+var StructArrayType = require('../util/struct_array');
+
+module.exports = draw;
+
+function draw(painter, source, layer, coords) {
+    if (layer.paint['fill-opacity'] === 0) return;
+    var gl = painter.gl;
+    gl.disable(gl.STENCIL_TEST);
+    painter.depthMask(true);
+
+    // Create a new texture to which to render the extrusion layer. This approach
+    // allows us to adjust opacity on a per-layer basis (eliminating the interior
+    // walls per-feature opacity problem)
+    var texture = new ExtrusionTexture(gl, painter, layer);
+    texture.bindFramebuffer();
+
+    gl.clearStencil(0x80);
+    gl.stencilMask(0xFF);
+    gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
+    gl.stencilMask(0x00);
```

Comment:
> testing for the clipping masks made me realize I forgot to flip the depthMask back off after this draw, which was causing fills to z-fight if above an extrusion layer — fixed 😅

nope I was totally wrong, this broke other fills and also is https://github.com/mapbox/mapbox-gl-js/issues/3320